File: selenium_tests_node/pages/registerPage.js
const BasePage = require('./basePage');

class RegisterPage extends BasePage {
    constructor(driver) {
        super(driver);
        this.TITLE = "Create Account";
        this.SUBTITLE = "Join the elite arena of competitive coders.";
        this.FULL_NAME_HINT = "John Doe";
        this.EMAIL_HINT = "dev@duel.io";
        this.COLLEGE_HINT = "Start typing your institute...";
        this.PASSWORD_HINT = "••••••••";
        this.REGISTER_BTN = "REGISTER";
        this.LOGIN_LINK = "Login";
    }

    async isOnRegisterPage(timeout = 15000) {
        return await this.isTextVisible(this.TITLE, timeout);
    }

    async enterFullName(name) {
        await this.enterTextByHint(this.FULL_NAME_HINT, name);
    }

    async enterEmail(email) {
        await this.enterTextByHint(this.EMAIL_HINT, email);
    }

    async enterCollege(college) {
        await this.enterTextByHint(this.COLLEGE_HINT, college);
    }

    async enterPassword(password) {
        // Find all password fields and fill the first one
        const fields = await this.driver.findElements(require('selenium-webdriver').By.xpath(`//*[@aria-label='${this.PASSWORD_HINT}'] | //input[@placeholder='${this.PASSWORD_HINT}']`));
        if (fields.length > 0) {
            await fields[0].click();
            await fields[0].clear();
            await fields[0].sendKeys(password);
        } else {
             // fallback via base method
             await this.enterTextByHint(this.PASSWORD_HINT, password);
        }
    }

    async enterConfirmPassword(password) {
        // Find all password fields and fill the last one
        const fields = await this.driver.findElements(require('selenium-webdriver').By.xpath(`//*[@aria-label='${this.PASSWORD_HINT}'] | //input[@placeholder='${this.PASSWORD_HINT}']`));
        if (fields.length > 1) {
            const lastField = fields[fields.length - 1];
            await lastField.click();
            await lastField.clear();
            await lastField.sendKeys(password);
        }
    }

    async clickRegister() {
        await this.clickByText(this.REGISTER_BTN);
    }

    async clickLoginLink() {
        await this.clickByText(this.LOGIN_LINK);
    }

    async register(name, email, college, password) {
        await this.enterFullName(name);
        await this.enterEmail(email);
        await this.enterCollege(college);
        await this.enterPassword(password);
        await this.enterConfirmPassword(password);
        await this.clickRegister();
    }

    async isMismatchErrorVisible(timeout = 8000) {
        return await this.isTextVisible("do not match", timeout);
    }

    async isShortPasswordErrorVisible(timeout = 8000) {
        return await this.isTextVisible("at least 6", timeout);
    }

    async isEmptyFieldsErrorVisible(timeout = 8000) {
        return await this.isTextVisible("fill in all fields", timeout);
    }

    async isSuccessVisible(timeout = 15000) {
        return await this.isTextVisible("Account created", timeout);
    }
}

module.exports = RegisterPage;

File: appium_tests/pages/register_page.py
"""
register_page.py – Registration screen page object
====================================================
Matches the RegisterScreen widget in devduel_mobile.
"""
import time
from appium.webdriver.common.appiumby import AppiumBy
from .base_page import BasePage


class RegisterPage(BasePage):
    """Page object for /register route."""

    TITLE          = "Create Account"
    SUBTITLE       = "Join the elite arena of competitive coders."
    FULL_NAME_HINT = "John Doe"
    EMAIL_HINT     = "dev@duel.io"
    COLLEGE_HINT   = "Start typing your institute..."
    PASSWORD_HINT  = "••••••••"
    REGISTER_BTN   = "REGISTER"
    LOGIN_LINK     = "Login"

    # ── Presence check ────────────────────────────────────────────────────────

    def is_on_register_page(self, timeout: int = 15) -> bool:
        return self.is_text_visible(self.TITLE, timeout)

    # ── Field helpers ─────────────────────────────────────────────────────────

    def enter_full_name(self, name: str):
        f = self.driver.find_element(AppiumBy.XPATH, "//*[@hint='John Doe']")
        f.click()
        f.clear(); f.send_keys(name)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass
        time.sleep(0.5)

    def enter_email(self, email: str):
        f = self.driver.find_element(AppiumBy.XPATH, "//*[@hint='dev@duel.io']")
        f.click()
        f.clear(); f.send_keys(email)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass
        time.sleep(0.5)

    def enter_college(self, college: str):
        f = self.driver.find_element(AppiumBy.XPATH, "//*[@hint='Start typing your institute...']")
        f.click()
        f.clear(); f.send_keys(college)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass
        time.sleep(0.5)

    def enter_password(self, password: str):
        fields = self.driver.find_elements(AppiumBy.XPATH, "//*[@hint='••••••••']")
        f = fields[0]
        f.click()
        f.clear(); f.send_keys(password)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass
        time.sleep(0.5)

    def enter_confirm_password(self, password: str):
        fields = self.driver.find_elements(AppiumBy.XPATH, "//*[@hint='••••••••']")
        f = fields[-1]
        f.click()
        f.clear(); f.send_keys(password)
        try:
            self.driver.hide_keyboard()
        except Exception:
            pass
        time.sleep(0.5)

    def tap_register(self):
        self.tap_by_text(self.REGISTER_BTN)

    def tap_login_link(self):
        self.tap_by_text(self.LOGIN_LINK)

    # ── Compound action ───────────────────────────────────────────────────────

    def register(self, name: str, email: str, college: str, password: str):
        self.enter_full_name(name)
        self.enter_email(email)
        self.enter_college(college)
        self.enter_password(password)
        self.enter_confirm_password(password)
        self.tap_register()

    # ── Validation ────────────────────────────────────────────────────────────

    def is_mismatch_error_visible(self, timeout: int = 8) -> bool:
        return self.is_text_contains_visible("do not match", timeout)

    def is_short_password_error_visible(self, timeout: int = 8) -> bool:
        return self.is_text_contains_visible("at least 6", timeout)

    def is_empty_fields_error_visible(self, timeout: int = 8) -> bool:
        return self.is_text_contains_visible("fill in all fields", timeout)

    def is_success_snackbar_visible(self, timeout: int = 15) -> bool:
        return self.is_text_contains_visible("Account created", timeout)
